Use ICU on Windows when available

safern opened the pull request:
Contributes towards: https://github.com/dotnet/runtime/issues/826
Fixes: https://github.com/dotnet/runtime/issues/4673

Mono is still defaulted to use NLS because we don't produce a shim in the libraries Windows build. Once the shim is linked into the mono runtime I'll update the mono code to use ICU when available. (cc: @marek-safar)

FYI: @GrabYourPitchforks @ericstj

safern committed:
Use ICU when available on Windows

safern committed:
Fix System.Globalization tests

safern committed:
Fix System.Globalization.Extensions tests

safern committed:
Fix System.Runtime tests

safern committed:
Add System.Globalization.Nls.Tests to test UseNls runtime switch

safern committed:
Add System.Globalization.Extensions.Nls.Tests to force them to run on NLS

safern committed:
Add System.Runtime.Nls.Tests.csproj to test Nls behavior

safern committed:
Merge remote-tracking branch 'dotnet/master' into GlobalizationIcuManagedImplementation

safern committed:
Fix left over from GlobalizationTestMode

jkotas commented on src/coreclr/src/System.Private.CoreLib/src/System/Globalization/GlobalizationMode.Unix.cs:
```diff
@@ -6,6 +6,8 @@ namespace System.Globalization
 {
     internal static partial class GlobalizationMode
     {
+        internal static bool UseIcu => true;
```
```suggestion
        internal const bool UseIcu = true;
```

This will allow C# compiler to strip the Icu code.

safern replied:
I believe it is also stripped as it is, but the change makes sense.

GrabYourPitchforks replied:
I tried making a similar change a while back for an unrelated code path and it ended up causing a bunch of "Unreachable code detected" build warnings. So be prepared to scatter a bunch of pragmas throughout the managed globalization call paths if we go this route.

safern replied:
Thanks. I'll try changing it and see how it goes 🤞

marek-safar replied:
It will produce unreachable warnings

jkotas replied:
It would be unfortunate to ship Windows-specific stuff on Unix. It would either disable the warnings (e.g. even for the whole file); or resort to ifdefs.

jkotas replied:
Once we get https://github.com/dotnet/runtime/issues/31785 , it can be a property and the linker can dead-code eliminate the Windows-specific code on Unix. So having an imperfect workaround - like disabling the warnings for the whole file - may be acceptable for now.

marek-safar replied:
#31785 is actually not required for this to work as the property is constant in the implementation. The linker will do the dead-code elimination correctly today

jkotas replied:
We do not have this enabled when building CoreLib today. What needs to happen to enable it? Can we do this before this change?

marek-safar replied:
I think it's if SPC is processed by linker during build (it's default setting on linker side)

safern replied:
I did this test before putting the PR and it stripped the Windows implementation on Unix correctly. I'll do another smoke test today to make sure and share screenshots and more detailed information but I believe @marek-safar is correct here.

safern replied:
I just validated this and I can see that it is actually being trimmed correctly. Here is the result from the pre-trim output: 
![image](https://user-images.githubusercontent.com/22899328/78832379-00995b00-79a0-11ea-9886-28add736d811.png)
![image](https://user-images.githubusercontent.com/22899328/78832425-14dd5800-79a0-11ea-9c2b-30a339041e6a.png)

Here is the result of searching for Nls on ILSpy on the final System.Private.CoreLib which is running the ILTrimmer:
![image](https://user-images.githubusercontent.com/22899328/78832159-a4ced200-799f-11ea-8155-8775000782c7.png)

![image](https://user-images.githubusercontent.com/22899328/78832228-bc0dbf80-799f-11ea-8938-52ab55d25812.png)

It might be worth adding a test into System.Runtime.Tests that tries and gets some of the windows types and methods to validate the trimmer is trimming and protect from any regression there. Thoughts?

Are we OK with leaving as a property?

jkotas replied:
> Are we OK with leaving as a property?

Sounds good.

I got confused by the calls to `Globalization.UseIcu` still being there. I expected them to disappear as well. @marek-safar Is there a reason for not stripping them?

marek-safar replied:
Yes, it's a linker and the goal was to remove dependencies and not do inlining which has significantly more complex rules, breaks debugging, etc. We could add another pass for limited inlining cases but I'm not sure it's worth it as existing JIT/AOT engines can handle this.

safern replied:
@marek-safar it seems that if I change `UseIcu` in Linux to be `UseIcu { get; } = GetUseIcuMode();` and then `GetUseIcuMode() => true;` the trimmer doesn't strip out any code, is that a known limitation?

jkotas replied:
That is expected. It is introducing non-trivial data flow that the linker is not able to crack.

jkotas commented on src/coreclr/src/System.Private.CoreLib/src/System/Globalization/GlobalizationMode.Windows.cs:
```diff
@@ -6,9 +6,27 @@ namespace System.Globalization
 {
     internal static partial class GlobalizationMode
     {
+        internal static bool UseIcu { get; private set; } = GetUseIcuMode();
```
```suggestion
        internal static bool UseIcu { get; } = GetUseIcuMode();
```
I believe having a setter on this will prevent the field from being readonly, that in turn prevents optimizations.

safern replied:
Ah right, that was a leftover from the initial way I had implemented this.

jkotas commented on src/libraries/Native/Unix/Common/pal_atomic.h:
```diff
@@ -9,11 +9,14 @@
+    return __atomic_compare_exchange_n(dest, comparand, exchange, false, __ATOMIC_SEQ_CST, __ATOMIC_SEQ_CST);
 #elif defined(TARGET_WINDOWS)
-    return InterlockedCompareExchangePointer(dest, exchange, comparand) == comparand;
+    void *value = *((void**)comparand);
```
This looks odd. I think the meaning of `pal_atomic_cas_ptr` arguments should match  `InterlockedCompareExchangePointer` .

safern replied:
The problem here is that the linux API compares the contents of the ptr and the windows `InterlockedCompareExchangePointer` compares the values. So when calling this API in pal_collation.c we need to do:

```c
UCollator* pNull = NULL;

pal_atomic_cas_ptr(..., ..., &pNull);
```

by passing the address of pNull so that the linux implementation actually compares against NULL. We would need to #ifdef the callsite. Do you have a suggestion how to improve that?

jkotas replied:
You can take the address of the argument in the Unix version of the helper.

safern replied:
right, but then I would need to #ifdef the callsite no?

jkotas replied:
No. The callsite would always use the `InterlockedCompareExchangePointer` style of arguments.

safern replied:
> You can take the address of the argument in the Unix version of the helper.

Yeah I see what you're saying. I think I tried that, but since we would be passing `pNull` which is a `NULL` ptr, can we actually get the address of that?

jkotas replied:
Yes, you should be able to get address of a variable holding null just fine.

safern replied:
Ok, I'll try it out.

tarekgh replied:
looks I am the one confused Santi with this :-)

---
In reply to: [405146796](https://github.com/dotnet/runtime/pull/34645#discussion_r405146796) [](ancestors = 405146796)

jkotas commented on src/mono/netcore/System.Private.CoreLib/src/System/Globalization/GlobalizationMode.Windows.Mono.cs:
```diff
@@ -6,9 +6,11 @@ namespace System.Globalization
 {
     internal partial class GlobalizationMode
     {
+        internal static bool UseIcu => false;
```
Change it to const field - similar to CoreCLR.

jkotas commented on src/coreclr/src/System.Private.CoreLib/src/System/Globalization/GlobalizationMode.Windows.cs:
```diff
@@ -6,9 +6,27 @@ namespace System.Globalization
+        private static bool GetUseIcuMode()
+        {
+            bool useNlsValue = GetUseNlsSwitchValue();
+            if (!Invariant)
```
This is fragile pattern - it only worth if the Invariant is initialized before IcuMode.

safern replied:
Wouldn't this force Invariant to be initialized? So you suggest instead calling `GetInvariantSwitchValue` here?

jkotas replied:
All static fields are initialized in the same static constructor. Accessing the static fields from the static constructors works only if the static field was initialized already.

The happens to work fine because GlobalizationMode.cs is before GlobalizationMode.Windows.cs in the .csproj file; but depending on this order is a fragile pattern that often leads to bugs.

jkotas replied:
I do not think we would want to call `GetInvariantSwitchValue` twice. I would put both the properties into the same file, next to each other, with a comment that the order is important.

safern replied:
Ok, I'll refactor this as suggested. Thanks.

jkotas commented on src/coreclr/src/System.Private.CoreLib/src/System/Globalization/GlobalizationMode.Windows.cs:
```diff
@@ -6,9 +6,27 @@ namespace System.Globalization
+
+        private static bool GetUseIcuMode()
+        {
+            bool useNlsValue = GetUseNlsSwitchValue();
```
Why is this before the `Invariant` check?

safern replied:
It was to return !useNlsValue... but I don't think it makes sense, if Invariant is set, it shouldn't matter what the value of `UseIcu` is.

tarekgh replied:
yes I have the same question as Jan.

---
In reply to: [405038481](https://github.com/dotnet/runtime/pull/34645#discussion_r405038481) [](ancestors = 405038481)

GrabYourPitchforks commented:
FYI I recently changed some of these files in https://github.com/dotnet/runtime/pull/34616. There will be conflicts until you merge in origin/master.

safern commented:
> FYI I recently changed some of these files in #34616. There will be conflicts until you merge in origin/master.

Thanks, will do so.

safern committed:
Merge remote-tracking branch 'dotnet/master' into GlobalizationIcuManagedImplementation

GrabYourPitchforks commented on src/libraries/System.Private.CoreLib/src/System/Globalization/JapaneseCalendar.Nls.cs:
```diff
@@ -114,6 +117,13 @@ public partial class JapaneseCalendar : Calendar
+        // the reason to have it is to simplify the build
+        // this way we avoid having to include RegistryKey
+        // and all it's windows PInvokes.
+        private static EraInfo[]? NlsGetJapaneseEras() => null;
```
So then should the implementation instead be as below?

```cs
private static EraInfo[]? NlsGetJapaneseEras()
{
    Debug.Fail("Should never be called on non-Windows platforms.");
    throw new PlatformNotSupportedException();
}
```

safern replied:
Yeah, good call.

GrabYourPitchforks reviewed (approved):
Nothing terribly breathtaking in my feedback. Feel free to ignore nits if you think they're not worthwhile. Otherwise most of the feedback is strengthening corner cases and future-proofing these code paths.

I'm so glad to see this feature coming together! :D

safern commented on src/libraries/Common/tests/CoreFx.Private.TestUtilities/System/PlatformDetection.cs:
```diff
@@ -150,6 +151,35 @@ public static string GetDistroVersionString()
+        public static bool ShouldUseIcu => ICUVersion > new Version(0,0,0,0);
+        public static bool ShouldUseNls => !ShouldUseIcu;
+
+        private static Version GetICUVersion()
```
> Or does GlobalizationNative_GetICUVersion() also work (and return 0) on Win7/Win8?

It will return 0 if ICU is not loaded, that's the reason why I'm adding the `u_getVersion_ptr != NULL` check in `GlobalizationNative_GetICUVersion`. That is how we're deciding whether we should run tests on NLS or ICU.

safern commented on src/libraries/Common/tests/Tests/System/StringTests.cs:
```diff
@@ -2616,7 +2618,7 @@ public static void Format_Invalid()
         }
 
-        [Theory]
+        [ConditionalTheory]
```
Yes, this was exactly to be able to skip the test at runtime based on the input. Basically by wrapping it under a `ConditionalTheory` with no arguments it will not skip the test but will wrap the testcase with this class: https://github.com/dotnet/arcade/blob/master/src/Microsoft.DotNet.XUnitExtensions/src/SkippedTheoryTestCase.cs#L23 which sets the message bus to be an instance of `SkippedMessageBus` which checks if the exception thrown is a `SkipTestException`, https://github.com/dotnet/arcade/blob/master/src/Microsoft.DotNet.XUnitExtensions/src/SkippedTestMessageBus.cs#L38 -- if that is the case then that test case is treated as skipped instead of failing and will print the exception message as the skipped reason to the console.

safern replied:
Btw, it doesn't have to be a parameterless `ConditionalTheory`, but since in this case we want the test case to always run and just skip one inline data case, I did it that way.

GrabYourPitchforks replied:
Thanks for the context!

GrabYourPitchforks commented:
I also didn't quite see where we're linking against the ICU libraries on Windows ([see MSDN](https://docs.microsoft.com/en-us/windows/win32/intl/international-components-for-unicode--icu-)). Was that part of an earlier commit?

safern commented on src/libraries/System.Private.CoreLib/src/System/Globalization/CalendarData.Icu.cs:
```diff
@@ -31,7 +31,7 @@ internal enum CalendarDataType
     internal partial class CalendarData
     {
-        private bool LoadCalendarDataFromSystem(string localeName, CalendarId calendarId)
+        private bool IcuLoadCalendarDataFromSystem(string localeName, CalendarId calendarId)
```
Sounds reasonable.

safern commented:
> I also didn't quite see where we're linking against the ICU libraries on Windows (see MSDN). Was that part of an earlier commit?

Yeah, I first linked System.Globalization.Native into coreclr here: https://github.com/dotnet/runtime/commit/be3ebf6db06a374f32032a2755fa3beb59e892de

Then I added support to build System.Globalization.Native on Windows using Windows ICU here:
https://github.com/dotnet/runtime/commit/d07a30cd9a2e83c0edd9a4d8d554b372bdab6767

safern commented on src/libraries/System.Globalization/tests/DateTimeFormatInfo/DateTimeFormatInfoLongTimePattern.cs:
```diff
@@ -48,9 +48,8 @@ public void LongTimePattern_SetReadOnly_ThrowsInvalidOperationException()
-        [Fact]
-        [PlatformSpecific(TestPlatforms.AnyUnix)]
-        public void LongTimePattern_CheckReadingTimeFormatWithSingleQuotes()
+        [ConditionalFact(typeof(PlatformDetection), nameof(PlatformDetection.ShouldUseIcu))]
```
We have a lot of pretty common conditions across tests using `ConditionalFact`, I think adding an extra attribute for that might be a little bit of overhead, but we can consider it in the future if we see this appear more often, at the moment it doesn't look like too many.

GrabYourPitchforks commented:
The PR https://github.com/dotnet/runtime/pull/34656 updated almost all of the libraries' solution files, so you'll need to react to that as well.

safern commented:
> The PR #34656 updated almost all of the libraries' solution files, so you'll need to react to that as well.

Yeah I just noticed that 😢

stephentoub reviewed (commented):
Nice job slogging through all of this.

jkotas commented on src/libraries/System.Private.CoreLib/src/System/Globalization/Normalization.cs:
```diff
@@ -0,0 +1,26 @@
+    internal static partial class Normalization
+    {
+        internal static bool IsNormalized(string strInput, NormalizationForm normalizationForm)
+        {
```
Factor our the `if (GlobalizationMode.Invariant)` call into this method as well?

jkotas commented on src/libraries/System.Private.CoreLib/src/System/Globalization/IdnMapping.Nls.cs:
```diff
@@ -104,17 +104,6 @@ private unsafe string GetUnicodeCore(string asciiString, char* ascii, int count,
-            get
-            {
-                int flags =
-                    (AllowUnassigned ? Interop.Normaliz.IDN_ALLOW_UNASSIGNED : 0) |
```
I see that these System.Globalization.Native flags happen to have the same values as the Windows NLS flags that allowed you to unify this; but I think it makes the code less readable. I would keep the flags and this method duplicated.

safern replied:
The reason why I did this was to avoid declaring a `NlsFlags` and an `IcuFlags` property. I'm curious, why do you think it makes code less readable?

jkotas replied:
It is not obvious that this constant is actually defined by Windows SDK, with a different name (https://docs.microsoft.com/en-us/windows/win32/api/winnls/nf-winnls-idntoascii). The coding convention is to use the platform names for constants defined by the platform.

I think it would be fine to have both `NlsFlags` and `IcuFlags` properties.

safern replied:
Yeah, I already changed it, I was curious why you thought it was less readable.

tarekgh commented on src/coreclr/src/System.Private.CoreLib/src/System/Globalization/GlobalizationMode.cs:
```diff
@@ -8,16 +8,21 @@ internal static partial class GlobalizationMode
+            GetSwitchValue("System.Globalization.Invariant", "DOTNET_SYSTEM_GLOBALIZATION_INVARIANT");
+
+        internal static bool GetUseNlsSwitchValue() =>
+            GetSwitchValue("System.Globalization.UseNls", "DOTNET_SYSTEM_GLOBALIZATION_USENLS");
```
>DOTNET_SYSTEM_GLOBALIZATION_USENLS [](start = 59, length = 34)

the term NLS could be not popular. I suggest to name this UseWin32

jkotas replied:
We should stick with established names and invent our own. I do not see Win32 in any of the docs talking about NLS vs ICU differences.

In fact, ICU globalization APIs are under Win32 path in the docs [https://docs.microsoft.com/en-us/windows/win32/intl/international-components-for-unicode--icu-](https://docs.microsoft.com/en-us/windows/win32/intl/international-components-for-unicode--icu-)

tarekgh replied:
Looking one more time at the docs, I am seeing the NLS is well defined. I agree then staying with NLS 

[https://docs.microsoft.com/en-us/windows/win32/intl/national-language-support](https://docs.microsoft.com/en-us/windows/win32/intl/national-language-support)

---
In reply to: [405674417](https://github.com/dotnet/runtime/pull/34645#discussion_r405674417) [](ancestors = 405674417)

tarekgh commented on src/libraries/Common/tests/CoreFx.Private.TestUtilities/System/PlatformDetection.cs:
```diff
@@ -150,6 +151,35 @@ public static string GetDistroVersionString()
+            }
+            catch { }
+
+            return new Version(version & 0xFF,
```
although this is a good point but it is not going to affect anything as we don't depend on the version in anyway more than just checking if it is 0 or for debugging purpose to know which version is used there. fixing it the was that Levi suggest wouldn't hurt though.

---
In reply to: [405118826](https://github.com/dotnet/runtime/pull/34645#discussion_r405118826) [](ancestors = 405118826)

tarekgh commented on src/libraries/System.Private.CoreLib/src/System/Globalization/CompareInfo.cs:
```diff
@@ -153,7 +156,28 @@ public static unsafe bool IsSortable(string text)
+            }
+        }
+
+        private static unsafe bool IsSortableCore(char* pChar, int length)
```
>IsSortableCore(char* pChar, int length) [](start = 35, length = 39)

nit: make it the style 

IsSortableCore(char* pChar, int length) => GlobalizationMode.UseIcu ? IcuIsSortable(pChar, length) :
 NlsIsSortable(pChar, length);

tarekgh replied:
maybe you need to apply this to other similar methods.

---
In reply to: [405682662](https://github.com/dotnet/runtime/pull/34645#discussion_r405682662) [](ancestors = 405682662)

tarekgh commented on src/libraries/System.Private.CoreLib/src/System/Globalization/CultureInfo.Icu.cs:
```diff
@@ -8,7 +8,7 @@ namespace System.Globalization
     public partial class CultureInfo : IFormatProvider
     {
-        internal static CultureInfo GetUserDefaultCulture()
+        internal static CultureInfo IcuGetUserDefaultCulture()
```
>IcuGetUserDefaultCulture [](start = 36, length = 24)

we need to double check ICU is going to return the correct default locale. I expect so but we need to double check it.

safern replied:
I'll double check this later. Thanks.

tarekgh reviewed (approved):
Great job Santi!

safern committed:
Merge remote-tracking branch 'dotnet/master' into GlobalizationIcuManagedImplementation

safern committed:
PR Feedback

safern commented on src/libraries/System.Private.CoreLib/src/System/Globalization/CompareInfo.cs:
```diff
@@ -153,7 +156,28 @@ public static unsafe bool IsSortable(string text)
+            }
+            else
+            {
+                _sortHandle = NlsGetSortHandle(_sortName);
```
Unfortunately, `NlsGetSortHandle` is used in `TextInfo` as a static method. What I think I can do is add an extra layer which is called from here which matches the `IcuInitSortHandle` signature and then `NlsInitSortHandle` calls down into `NlsGetSortHandle`.

mjsabby commented:
Does this mean that ICU will be used on Windows by default?

safern commented:
> Does this mean that ICU will be used on Windows by default?

Yes on Windows10 19H1+. You will be able to turn it off by setting `DOTNET_SYSTEM_GLOBALIZATION_USENLS=1` env var or by setting a runtime switch: `System.Globalization.UseNls: true`.

safern commented:
I've addressed PR Feedback. Still need to work on fixing the mono builds and looking at other tests failing due to this change. Will work on that and update the PR.

jkotas commented on src/coreclr/src/System.Private.CoreLib/src/System/Globalization/GlobalizationMode.cs:
```diff
@@ -6,18 +6,32 @@ namespace System.Globalization
         internal static bool Invariant { get; } = GetGlobalizationInvariantMode();
 
-        // GetInvariantSwitchValue calls CLRConfig first to detect if the switch is defined in the config file.
+#if TARGET_WINDOWS
```
We have .Unix.cs and .Windows.cs files for this. Why not move this to those files?

jkotas replied:
Ah ok, this is because of the feedback I had about the order.

safern replied:
So that the properties were in the same file for the initialization order. The other way would be to move Invariant and duplicate it to the `Unix/Windows` files. Is that what you'd like.

safern replied:
> Ah ok, this is because of the feedback I had about the order.

Yeah, but I can instead duplicate Invariant to avoid the #ifdef -- would you prefer that?

jkotas replied:
Yeah, I think avoiding the ifdef and duplicating a few lines between Windows/Unix would look better.

safern replied:
Sounds good. I'll do that.

jkotas commented on src/coreclr/src/System.Private.CoreLib/src/System/Globalization/GlobalizationMode.cs:
```diff
@@ -6,18 +6,32 @@ namespace System.Globalization
+        internal static bool GetInvariantSwitchValue() =>
+            GetSwitchValue("System.Globalization.Invariant", "DOTNET_SYSTEM_GLOBALIZATION_INVARIANT");
+
+        internal static bool GetUseNlsSwitchValue() =>
```
This can be in the .Windows.cs file too.

safern replied:
Make sense.

jkotas replied:
And maybe used directly, without the wrapper.

jkotas commented on src/coreclr/src/System.Private.CoreLib/src/System/Globalization/GlobalizationMode.cs:
```diff
@@ -6,18 +6,32 @@ namespace System.Globalization
     {
+        // Order of these properties matter because GetUseIcuMode is dependent on Invariant.
+        // So we need Invariant to be initialized first.
         internal static bool Invariant { get; } = GetGlobalizationInvariantMode();
```
```suggestion
        internal static bool Invariant { get; }
            = GetSwitchValue("System.Globalization.Invariant", "DOTNET_SYSTEM_GLOBALIZATION_INVARIANT");
```

This goes through two layers of wrappers. Why not just call it directly?

safern replied:
Not for the .Unix.cs file. The .Unix.cs file has some logic before calling `GetInvariantSwitchValue`.

safern replied:
I see what you’re saying now. I’ll fix it

jkotas replied:
Ok, this should be easier to see https://github.com/dotnet/runtime/pull/34645#discussion_r405957915

jkotas commented on src/libraries/Native/Unix/System.Globalization.Native/pal_icushim.c:
```diff
@@ -374,7 +374,10 @@ int32_t GlobalizationNative_LoadICU()
 {
-    int32_t version;
-    u_getVersion((uint8_t *) &version);
+    int32_t version = 0;
```
This would avoid the dirty cast to `uint8_t *`, and also make the IsLittleEndian check unnecessary at the place where this is called:

```
UVersionInfo versionInfo;

if (u_getVersion_ptr == NULL)
    return 0;

u_getVersion(&versionInfo);

return (versionInfo[0] << 24) + (versionInfo[1] << 16) + (versionInfo[2] << 8) + versionInfo[3];
```

safern replied:
Nice 😄

safern replied:
It seems like even after this we either need versionInfo to be a uint_8* or cast it. I'm getting:

```
'void (uint8_t *)': cannot convert argument 1 from 'UVersionInfo (*)
' to 'uint8_t *'
```

is that maybe because of the way we define the _ptr for the function with the macros and UVersionInfo is getting replaced with uint8_t?

This is what icu.h has:
```c
typedef uint8_t UVersionInfo[U_MAX_VERSION_LENGTH];
u_getVersion(UVersionInfo versionArray);
```

I guess I'm going to declare my array using `[U_MAX_VERSION_LENGTH];` instead.

jkotas replied:
I think replacing `u_getVersion(&versionInfo);` -> `u_getVersion(versionInfo);` in my suggestion should fix this.

safern replied:
Ah right, I think that should fix it cause it is already a ptr. I'll try it out.

jkotas commented on src/libraries/Common/tests/TestUtilities/System/PlatformDetection.cs:
```diff
@@ -150,6 +151,44 @@ public static string GetDistroVersionString()
+        private static Lazy<Version> m_icuVersion = new Lazy<Version>(GetICUVersion);
+        public static Version ICUVersion => m_icuVersion.Value;
+
+        public static bool ShouldUseIcu => ICUVersion > new Version(0,0,0,0);
```
Nit: No other property on PlatformDetection is called ShouldXXX.

Would `IsIcuGlobalization` be a better name for this?

safern replied:
Yeah, that's a good suggestion.

jkotas commented on src/coreclr/src/System.Private.CoreLib/src/System/Globalization/GlobalizationMode.cs:
```diff
@@ -6,18 +6,32 @@ namespace System.Globalization
+            GetSwitchValue("System.Globalization.Invariant", "DOTNET_SYSTEM_GLOBALIZATION_INVARIANT");
+
+        internal static bool GetUseNlsSwitchValue() =>
+            GetSwitchValue("System.Globalization.UseNls", "DOTNET_SYSTEM_GLOBALIZATION_USENLS");
```
Is there a reason why the config switch is called UseNls, and then we invert it and have a property called UseIcu ? It would be nice to be consistent.

safern replied:
The reason why I chose that name was to tell our customers that Icu is the default, so if you want to use nls set `UseNls` explicitly to true and also to make sure the name has NLS, so that it is clear is a Windows thing. You think it would be better to name this UseIcu and then customers would turn it off instead of on?

The reason why the property is `UseIcu` was to not invert conditions on the code and so the ICU code path was always the first code path in the condition since that is the default behavior. I think having customers turn it off might make sense.

jkotas replied:
I think using UseNls for to config switch makes sense.

I do not think it matters which code path is the first. It is fairly common for the special less common path to be first.

safern replied:
So you think it would be good to rename the property to `UseNls`? Ok, I can change that if you'd like.

jkotas replied:
Yes, I would make the config switch and the property to have the same name.

marek-safar replied:
/cc @vitek-karas

vitek-karas replied:
/cc @eerhardt - basically another feature switch

jkotas commented on src/coreclr/src/System.Private.CoreLib/src/System/Globalization/GlobalizationMode.Windows.cs:
```diff
@@ -10,5 +10,18 @@ private static bool GetGlobalizationInvariantMode()
+                }
+            }
+
+            return true;
```
```suggestion
            return false;
```

jkotas commented on src/coreclr/src/System.Private.CoreLib/src/System/Globalization/GlobalizationMode.Windows.cs:
```diff
@@ -10,5 +10,18 @@ private static bool GetGlobalizationInvariantMode()
+                }
+            }
+
+            return true;
```
This will return true for InvariantMode. I would expect it to return false.

(It should not matter, but it is confusing to UseIcu to be true when we are not using ICU.)

safern replied:
Yeah that's true, I just used `true` as an arbitrary value as it shouldn't matter, but you made a good point.

safern committed:
Merge remote-tracking branch 'dotnet/master' into GlobalizationIcuManagedImplementation

safern committed:
PR Feedback 2

safern committed:
Fix mono build

safern committed:
Fix failing tests

safern committed:
Fix typo

jkotas commented on src/coreclr/src/System.Private.CoreLib/src/System/Globalization/GlobalizationMode.cs:
```diff
@@ -6,18 +6,18 @@ namespace System.Globalization
     internal static partial class GlobalizationMode
     {
-        internal static bool Invariant { get; } = GetGlobalizationInvariantMode();
+        internal static bool GetInvariantSwitchValue() =>
```
```suggestion
        private static bool GetInvariantSwitchValue() =>
```

jkotas commented on src/coreclr/src/System.Private.CoreLib/src/System/Globalization/GlobalizationMode.Windows.cs:
```diff
@@ -6,9 +6,25 @@ namespace System.Globalization
+        private static bool GetUseNlsMode()
         {
-            return GetInvariantSwitchValue();
+            bool retVal = true;
+            if (!Invariant)
+            {
+                if (!GetSwitchValue("System.Globalization.UseNls", "DOTNET_SYSTEM_GLOBALIZATION_USENLS") &&
+                                    Interop.Globalization.LoadICU() != 0)
+                {
+                    retVal = false;
+                }
+            }
+
+            return retVal;
```
```suggestion
            if (Invariant)
                return false;
                
            return GetSwitchValue("System.Globalization.UseNls", "DOTNET_SYSTEM_GLOBALIZATION_USENLS") ||
                                    Interop.Globalization.LoadICU() == 0;
```

safern replied:
Good suggestion. Thanks.

safern replied:
I think we can remove the intermediate `GetUseNlsMode` method and do:
```cs
internal static bool UseNls { get; } = !Invariant && (GetSwitchValue("System.Globalization.UseNls", "DOTNET_SYSTEM_GLOBALIZATION_USENLS") ||
                                         Interop.Globalization.LoadICU() == 0);
```

Will instead do that.

jkotas commented on src/libraries/System.Private.CoreLib/src/System/Globalization/CalendarData.Icu.cs:
```diff
@@ -450,7 +456,7 @@ private static unsafe void EnumCalendarInfoCallback(string calendarString, IntPt
         }
 
-        private struct EnumCalendarsData
+        public struct IcuEnumCalendarsData
```
```suggestion
        private struct IcuEnumCalendarsData
```
?

I do not see why it had to change to public.

safern replied:
It seems like this was when renaming the member I accidentally changed it to public.

safern committed:
Fix bad merge in projitems for tvOS

safern committed:
PR Feedback 3

jkotas commented on src/coreclr/src/System.Private.CoreLib/src/System/Globalization/GlobalizationMode.Windows.cs:
```diff
@@ -6,9 +6,12 @@ namespace System.Globalization
+        internal static bool Invariant { get; } = GetInvariantSwitchValue();
+
+        internal static bool UseNls { get; } = !Invariant &&
+            (GetSwitchValue("System.Globalization.UseNls", "DOTNET_SYSTEM_GLOBALIZATION_USENLS") || Interop.Globalization.LoadICU() == 0);
```
```suggestion
            (GetSwitchValue("System.Globalization.UseNls", "DOTNET_SYSTEM_GLOBALIZATION_USENLS") ||
                Interop.Globalization.LoadICU() == 0);
```

safern committed:
PR Feedback 4

mjsabby commented:
Another question. We use the sysglobl APIs to create additional cultures and reparent them to existing ones in most cases. This afaik ultimately are just additions to the registry. Will such a feature will be available to ICU?

safern committed:
Merge remote-tracking branch 'dotnet/master' into GlobalizationIcuManagedImplementation

tarekgh commented:
@mjsabby 

> Another question. We use the sysglobl APIs to create additional cultures and reparent them to existing ones in most cases. This afaik ultimately are just additions to the registry. Will such a feature will be available to ICU?

Unfortunately, we'll be limited to what the underlying ICU library is providing to us. I checked with the Windows team and learned custom locales are not supported by ICU when running on Windows. It would be nice if you can tell why you are using such custom locales? 
If it is really very important to you then you may use the ICU app local feature which will allow you to load the app local version of the ICU and not the system one. And you'll need to build and install your own ICU customized locales data file. Although this can be non-trivial it is not impossible to do.

CC @jefgen @ShawnSteele

mjsabby commented:
@tarekgh For business and legal reasons we need to define few markets differently than Windows. So we definitely need to use the app local feature.

The question is how much more complicated is it that than the current process which is calling sysglobl APIs.

Is the app-local feature documentation available somewhere?

tarekgh commented:
> For business and legal reasons we need to define few markets differently than Windows.

Is this mean you need to update some locale data but not adding any more locales?

> Is the app-local feature documentation available somewhere?

This is not implemented yet. we are going to implement it next couple of weeks. Basically app local feature is the app will carry its own copy of ICU. Using the tool https://github.com/unicode-org/icu/blob/master/docs/userguide/icu_data/buildtool.md I believe you can build your own customized copy of ICU data and use it. I didn't play with this tool before but I think it will give you full control over the locale data for your app. Also, you'll have control when you can pick a newer version of the ICU at the time you like.
